ms_teams_webclient.py

```diff
@@ -0,0 +1,153 @@
+
+authtoken = namedtuple('AuthToken', 'access_token, expired_at')
+activity = namedtuple('Activity', 'post_url, payload')
+
+reactions = {
+    'thumbsup': '👍'
+}
+
+class MSTeamsWebclient:
+    def __init__(self, app_id, app_password, emulator_mode):
+        self.__app_id = app_id
+        self.__app_password = app_password
+        self.__emulator_mode = emulator_mode
+        self.__token = None
+
+    def send_message(self, identifier, message):
+        member = self.__get_member_by_email(identifier)
+        conversation = self.__create_conversation(member['id'], identifier.extras)
+        identifier.extras['conversation'] = conversation
+        self.__send_direct_message(message, identifier.extras)
+
+    def get_member_by_id(self, member_id, extras):
+        response = requests.get(
+            f'{extras["service_url"]}/v3/conversations/{extras["conversation_id"]}/members/{member_id}',
+            headers=self.__get_default_headers()
+        )
+        if response.status_code == 404:
+            raise MemberNotFound(f"member not found using {id} id")
+        return response.json()
+    
+    def send_reply(self, response):
+        """Post response to callback url
+
+        Send a reply to URL indicated in serviceUrl from
+        Bot Framework's request object.
+
+        @param response: activity object
+        """
+        r = requests.post(
+            response.post_url,
+            data=json.dumps(response.payload),
+            headers=self.__get_default_headers()
+        )
+
+        r.raise_for_status()
+
+    def add_reaction(self, message, reaction):
+        emoji = reactions[reaction] if reactions[reaction] else ''
+        message.body = f'> **@{message.frm.nick}**: {message.body}\n\n{emoji}'
+        reply = self.build_reply(message)
+        self.send_reply(reply)
+
+    def get_token(self):
+        now = datetime.datetime.now()
+        if not self.__token or self.__token.expired_at <= now:
+            self.__token = self.__auth()
+        return self.__token.access_token
+    
+    def build_reply(self, message):
+        req = message.extras['conversation']
+        payload = {
+            'type': 'message',
+            'conversation': req.conversation,
+            'from': message.to.subject,
+            'recipient': message.frm.subject,
+            'replyToId': req.conversation_id,
+            'text': message.body,
+            'textFormat': 'markdown'
+        }
+        return activity(req.reply_url, payload)
+
+    def __get_member_by_email(self, identifier):
+        service_url = identifier.extras['service_url']
+        team_id = identifier.extras['team_id']
+        email = identifier.email
+
+        response = requests.get(
+            f'{service_url}/v3/conversations/{team_id}/members/{email}',
+            headers=self.__get_default_headers()
+        )
+        if response.status_code == 404:
+            raise MemberNotFound(f"member not found using {identifier.email} email")
```

Comment:
I believe that raise_for_status is much better than custom exception because it preserves response details
```suggestion
r.raise_for_status()
```

Reply:
We added `raise_for_status` in order to keep the stack trace, but we still have a custom Exception so we can better treat the error at the bot side.